pkg/activator/net/revision_backends.go

```diff
@@ -0,0 +1,414 @@
+	"knative.dev/serving/pkg/apis/serving"
+	servinglisters "knative.dev/serving/pkg/client/listers/serving/v1alpha1"
+	"knative.dev/serving/pkg/network"
+	"knative.dev/serving/pkg/network/prober"
+	"knative.dev/serving/pkg/queue"
+	"knative.dev/serving/pkg/reconciler"
+)
+
+// RevisionDestsUpdate contains the state of healthy l4 dests for talking to a revision and is the
+// primary output from the RevisionBackendsManager system. If a healthy ClusterIP is found then
+// ClusterIPDest will be set to non empty string and Dests will be nil. Otherwise Dests will be set
+// to a slice of healthy l4 dests for reaching the revision.
+type RevisionDestsUpdate struct {
+	Rev           types.NamespacedName
+	ClusterIPDest string
+	Dests         []string
+}
+
+const (
+	probePortName  string        = "http"
+	probeTimeout   time.Duration = 300 * time.Millisecond
+	probeFrequency time.Duration = 200 * time.Millisecond
+)
+
+// revisionWatcher watches the podIPs and ClusterIP of the service for a revision. It implements the logic
+// to supply RevisionDestsUpdate events on updateCh
+type revisionWatcher struct {
+	rev      types.NamespacedName
+	protocol networking.ProtocolType
+	updateCh chan<- *RevisionDestsUpdate
+	// Stores the podIPs we know of
+	dests []string
+	// Stores the state of dests (podIPs) we know about
+	healthStates map[string]bool
+	// Stores whether the service ClusterIP has been seen as healthy
+	clusterIPHealthy bool
+
+	transport     http.RoundTripper
+	destsChan     <-chan []string
+	serviceLister corev1listers.ServiceLister
+	logger        *zap.SugaredLogger
+}
+
+func newRevisionWatcher(rev types.NamespacedName, protocol networking.ProtocolType,
+	updateCh chan<- *RevisionDestsUpdate, destsChan <-chan []string,
+	transport http.RoundTripper, serviceLister corev1listers.ServiceLister,
+	logger *zap.SugaredLogger) *revisionWatcher {
+	return &revisionWatcher{
+		rev:           rev,
+		protocol:      protocol,
+		updateCh:      updateCh,
+		healthStates:  make(map[string]bool),
+		transport:     transport,
+		destsChan:     destsChan,
+		serviceLister: serviceLister,
+		logger:        logger,
+	}
+}
+
+func filterHealthyDests(dests map[string]bool) []string {
+	ret := make([]string, 0, len(dests))
+	for dest, healthy := range dests {
+		if healthy {
+			ret = append(ret, dest)
+		}
+	}
+	return ret
+}
+
+type destHealth struct {
+	dest   string
+	health bool
+}
+
+func (rw *revisionWatcher) revisionReady() bool {
+	return len(rw.dests) > 0
+}
+
+func (rw *revisionWatcher) getK8sPrivateService() (*corev1.Service, error) {
+	selector := labels.SelectorFromSet(map[string]string{
+		serving.RevisionLabelKey:  rw.rev.Name,
+		networking.ServiceTypeKey: string(networking.ServiceTypePrivate),
+	})
+	svcList, err := rw.serviceLister.Services(rw.rev.Namespace).List(selector)
+	if err != nil {
+		return nil, err
+	}
+
+	switch len(svcList) {
+	case 0:
+		return nil, fmt.Errorf("Found no private service for revision %q", rw.rev.String())
```

Comment:
```suggestion
		return nil, fmt.Errorf("found no private service for revision %q", rw.rev.String())
```

Reply:
Is wanting lowercase for start of an error message actually a thing? If so can we do this via a linter or something.

Reply:
Yes, it is standard Go practice (https://github.com/golang/go/wiki/CodeReviewComments#error-strings)

Reply:
It's also reported by linters IIRC, at least by `staticcheck`.